File: problems/696-count-binary-substrings/solution.java
class Solution {
    public int countBinarySubstrings(String s) {
        int[] groups = new int[s.length()];
        int i = 0, j = 0;
        while (i < s.length()) {
            char c = s.charAt(i);
            int count = 0;
            while (i < s.length() && s.charAt(i) == c) {
                i++;
                count++;
            }
            groups[j++] = count;
        }
        int ans = 0;
        for (int k = 1; k < j; k++) {
            ans += Math.min(groups[k - 1], groups[k]);
        }
        return ans;
    }
}

File: problems/696-count-binary-substrings/solution.cpp
class Solution {
public:
    int countBinarySubstrings(string s) {
        int count = 0;
        int prev = 0;
        int curr = 1;
        for (int i = 1; i < s.length(); i++) {
            if (s[i] == s[i - 1]) {
                curr++;
            } else {
                count += min(prev, curr);
                prev = curr;
                curr = 1;
            }
        }
        count += min(prev, curr);
        return count;
    }
};